Payment/Program.cs

```diff
@@ -1,2 +1,183 @@
+		}
+
+		var payment = PaymentService.Create(paymentType, price);
+
+		var response = payment.ProcessPayment(price);
+
+		Console.WriteLine(response.Message);
+	}
+}
+
+public class PaymentService
+{
+	public static IPaymentService Create(PaymentType type, decimal price)
+	{
+		switch (type)
+		{
+			case PaymentType.CreditCard:
+				{
+					var factory = new CreditCardFactory();
+					return factory.CreateInstance();
+				}
+			case PaymentType.CryptoCurrency:
+				{
+					var factory = new CryptoCurrencyFactory();
+					return factory.CreateInstance();
+				}
+			default:
+				return new UnsupportedPayment();
+		}
+	}
+}
+
+public interface IPaymentService
+{
+	Payment ProcessPayment(decimal price);
+}
+
+public sealed class Payment
+{
+	public decimal? Price { get; init; }
+	public string? Message { get; init; }
+}
+
+public enum PaymentType
+{
+	CreditCard,
+	CryptoCurrency,
+	Unsupported
+}
+
+
+public class CreditCard : IPaymentService
+{
+	public Payment ProcessPayment(decimal price)
+	{
+		var calc = new CreditCardCalculator();
+		var finalAmount = calc.CalculatePayment(price);
+
+		var response = new Payment
+		{
+			Message = $"Processed credit card payment of {finalAmount}.",
+			Price = finalAmount
+		};
+		return response;
+	}
+}
+
+public class CryptoCurrencyAdapter : IPaymentService
+{
+	private readonly CryptoCurrency _cryptoCurrency = new();
+
+	public Payment ProcessPayment(decimal price)
+	{
+		var response = new Payment
+		{
+			Message = $"Processed crypto currency payment of {price}.",
+			Price = _cryptoCurrency.PayByCrypto(price)
+		};
+		return response;
+	}
+}
+
+public interface IPaymentCalculator
+{
+	decimal CalculatePayment(decimal price);
+}
+
+public class CreditCardCalculator : IPaymentCalculator
+{
+	public decimal CalculatePayment(decimal price)
+	{
+		return price + price * (decimal)0.02;
+	}
+}
+
+public class UnsupportedPayment : IPaymentService
+{
+	public Payment ProcessPayment(decimal price)
+	{
+		return new Payment
+		{
+			Message = "Unsupported Payment"
+		};
+	}
+}
+
+
+public interface IPaymentServiceFactory
+{
+	IPaymentService CreateInstance();
+}
+
+public class CreditCardFactory : IPaymentServiceFactory
+{
+	public IPaymentService CreateInstance()
+	{
+		return new CreditCard();
+	}
+}
+
+public class CryptoCurrencyFactory : IPaymentServiceFactory
+{
+	public IPaymentService CreateInstance()
+	{
+		return new CryptoCurrencyAdapter();
+	}
+}
+
+// External library class (inaccessible for modification)
+//public class CryptoCurrency
+//{
+//	public decimal PayByCrypto(decimal price)
+//	{
+//		return price;
+//	}
+//}
+
+public sealed class CryptoCurrency
+{
+	private const decimal ConversionRate = 0.0005m; 
+	private const decimal TransactionFee = 0.02m; 
+
+	public decimal PayByCrypto(decimal price)
+	{
+		var cryptoAmount = ConvertToCrypto(price);
+		var totalAmount = ApplyTransactionFee(cryptoAmount);
+
+		Console.WriteLine($"پرداخت کریپتو به مبلغ {totalAmount} در حال پردازش است.");
```

Comment:
The output for CryptoCurrency states that no adjustment was made, yet it processes the payment. Similarly, other payment methods don't consistently indicate whether adjustments were made.